### PR TITLE
Hw2

### PR DESCRIPTION
**Pull request by stsepchanka**

### CODE REVIEW

File: src/middleware/user.middleware.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
+import { Request, Response } from "express";
+import { StatusCodes } from "http-status-codes";
+
+import { UserService } from "../services";
+
+export class UserMiddleware {
```
Comment (kuzanatoliy) [inline src/middleware/user.middleware.ts]: I would use UserController name

File: src/middleware/user.middleware.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  constructor(private readonly userService: UserService) {}
+
+  autoSuggest() {
+    return (req: Request, res: Response): void => {
```
Comment (kuzanatoliy) [inline src/middleware/user.middleware.ts]: There is a reason to return function? I would use the autoSuggest method as middleware
    autoSuggest(req: Request, res: Response): void {
      try {
        const loginSubstring = req.query.loginSubstring as string;
        const limit = parseInt(req.query.limit as string);
        const users = this.userService.getAutoSuggestUsers(
          loginSubstring,
          limit
        );
        res.status(StatusCodes.OK).json(users);
      } catch (err) {
        res.status(StatusCodes.BAD_REQUEST).send(err);
      }
    }

File: src/middleware/user.middleware.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  updateUser() {
+    return (req: Request, res: Response): void => {
+      try {
+        const userId = req.params.id;
+        const user = this.userService.updateUser(userId, req.body);
+        res.status(StatusCodes.OK).json(user);
+      } catch (err) {
+        res.status(StatusCodes.NOT_FOUND).send(err);
```
Comment (kuzanatoliy) [inline src/middleware/user.middleware.ts]: I wouldn't send an error object here. Only status. res.sendStatus

File: src/models/user.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import * as Joi from "joi";
+
+export type User = {
```
Comment (kuzanatoliy) [inline src/models/user.ts]: In my opinion interface is better here and I also use I prefix in the interface case

File: src/repositories/user.repository.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    return this.users.filter((user) => !user.isDeleted);
+  }
+
+  getUserByID(userId: string): User {
+    const index = this.findUserIndex(userId);
+
+    return this.users[index];
+  }
+
+  addUser(user: User): User {
```
Comment (kuzanatoliy) [inline src/repositories/user.repository.ts]: Please split checking user and addicting functionality (Single responsiblility)

File: src/repositories/user.repository.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
+import { v4 as uuidv4 } from "uuid";
+
+import { User } from "../models";
+
+export class UserRepository {
```
Comment (kuzanatoliy) [inline src/repositories/user.repository.ts:5]: I'm a little confused. Why do you use the repository name? Looks as service

File: src/services/user.service.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+export class UserService {
+  constructor(private readonly userRepository: UserRepository) {}
+
+  getUsers(): User[] {
+    return this.userRepository.getUsers();
+  }
+
+  getUserByID(userId: string): User {
+    return this.userRepository.getUserByID(userId);
```
Comment (kuzanatoliy) [inline src/services/user.service.ts:12]: I'm not sure that repository is a necessary layer

File: src/middleware/user.middleware.ts (hunk 2 above)
Comment (stsepchanka) [reply src/middleware/user.middleware.ts]: Yes, I have to use 'return function' here to keep the context due to closure. Methods in UserMiddleware use 'this' and they lose the context if I send them to router without calling:
 `router.route("/auto-suggest").get(userMiddleware.autoSuggest)`

File: src/middleware/user.middleware.ts (hunk 1 above)
Comment (stsepchanka) [reply src/middleware/user.middleware.ts]: Renamed. Thank you.

File: src/middleware/user.middleware.ts (hunk 3 above)
Comment (stsepchanka) [reply src/middleware/user.middleware.ts]: Changed. Thank you.

File: src/models/user.ts (hunk 4 above)
Comment (stsepchanka) [reply src/models/user.ts]: Changed. Thank you.

File: src/repositories/user.repository.ts (hunk 5 above)
Comment (stsepchanka) [reply src/repositories/user.repository.ts]: One more method is added. Thank you.

File: src/repositories/user.repository.ts (hunk 6 above)
Comment (stsepchanka) [reply src/repositories/user.repository.ts:5]: The idea of using a Repository pattern is to create simple CRUD operations with data, to be able to change one DB to another (or in-memory data to DB). 
I think the service is responsible for more complicated logic (e.g. getAutoSuggestUsers) and for combining data from different repositories (for example, users and departments).

File: src/services/user.service.ts (hunk 7 above)
Comment (stsepchanka) [reply src/services/user.service.ts:12]: Probably, Repository is a little excessive in our case. If you don't mind I would leave it like it is and exclude it in the next homework.

File: src/middleware/user.middleware.ts (hunk 2 above)
Comment (kuzanatoliy) [reply src/middleware/user.middleware.ts]: You just can use arrow function

File: src/repositories/user.repository.ts (hunk 6 above)
Comment (kuzanatoliy) [reply src/repositories/user.repository.ts:5]: OK. But not needed here. When you will work with data base you will create DB Models

**Pull request merged**